R/finalsize_risk_grps.R

```diff
@@ -0,0 +1,120 @@
+#' work well.
+#' @param adapt_step Boolean, whether the solver step rate should be changed
+#' based on the solver error. Defaults to TRUE.
+#'
+#' @keywords epidemic model
+#' @return A vector of final sizes by demography group.
+#' @export
+#' @examples
+#' library(socialmixr)
+#' data(polymod)
+#' r0 <- 2.0
+#' contact_data <- contact_matrix(
+#'   polymod,
+#'   countries = "United Kingdom",
+#'   age.limits = c(0, 20, 40)
+#' )
+#' c_matrix <- t(contact_data$matrix)
+#' d_vector <- contact_data$participants$proportion
+#' # Scale contact matrix to demography
+#' c_matrix <- apply(
+#'   c_matrix, 1, function(r) r / d_vector
+#' )
+#' n_demo_grps <- length(d_vector)
+#' n_risk_grps <- 3
+#' # prepare p_susceptibility and susceptibility
+#' psusc <- matrix(
+#'   data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+#' )
+#' psusc <- t(apply(psusc, 1, \(x) x / sum(x)))
+#' susc <- matrix(
+#'   data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+#' )
+#' final_size_grps(
+#'   contact_matrix = r0 * c_matrix,
+#'   demography_vector = d_vector,
+#'   p_susceptibility = psusc,
+#'   susceptibility = susc
+#' )
+#'
+final_size_grps <- function(contact_matrix,
+                            demography_vector,
+                            p_susceptibility,
+                            susceptibility,
+                            iterations = 1000,
+                            tolerance = 1e-6,
+                            step_rate = 1.9,
+                            adapt_step = TRUE) {
+  # check arguments input
+  stopifnot(
+    "Error: contact matrix must have as many rows as demography groups" =
+      (nrow(contact_matrix) == length(demography_vector))
+  )
+  stopifnot(
+    "Error: p_susceptibility must have as many rows as demography groups" =
+      (nrow(p_susceptibility) == length(demography_vector))
+  )
+  stopifnot(
+    "Error: susceptibility must have as many rows as demography groups" =
+      (nrow(susceptibility) == length(demography_vector))
+  )
+  stopifnot(
+    "Error: susceptibility must have same dimensions as p_susceptibility" =
+      (all(dim(p_susceptibility) == dim(susceptibility)))
+  )
+  # p_susceptibility must have rowwise sums of 1.0 or nearly 1.0
+  stopifnot(
+    "Error: p_susceptibility rows must sum to 1.0" =
+      (
+        all(abs(rowSums(p_susceptibility) - 1) < 1e-6)
+      )
+  )
+
+  # prepare epi spread object
+  epi_spread_data <- epi_spread(
+    contact_matrix = contact_matrix,
+    demography_vector = demography_vector,
+    p_susceptibility = p_susceptibility,
+    susceptibility = susceptibility
+  )
+
+  # solve for final size
+  pi <- solve_final_size_iterative(
+    contact_matrix = epi_spread_data[["contact_matrix"]],
+    demography_vector = epi_spread_data[["demography_vector"]],
+    p_susceptibility = epi_spread_data[["p_susceptibility"]],
+    susceptibility = epi_spread_data[["susceptibility"]],
+    iterations = iterations,
+    tolerance = tolerance
```

Comment:
Function must also take `adapt_step` and `step_rate` arguments passed from wrapping function.

Reply:
Issue opened as #53 and fixed